### PR TITLE
LGA-785 - Adding Education and Discrimination categories to LAALAA

### PR DESCRIPTION
**Pull request by said-moj**
## What does this pull request do?

Adds  Education and Discrimination categories to LAALAA

## Any other changes that would benefit highlighting?

There seems to be a Integrity Constraint issue when inserting new categories on local and when running test. The cause is that the next available sequence for the id field is 1 when there are 13 records with ids ranging from 1-13. This why I have added reset_adviser_category_pkey method to update the sequence for the next available id before creating the new categories.

## Checklist

- [x] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        self.reset_adviser_category_pkey()
+        Category.objects.get_or_create(code="DISC", defaults={"civil": True})
+        Category.objects.get_or_create(code="EDU", defaults={"civil": True})
+
+    def reset_adviser_category_pkey(self):
+        # This is required because the id field is specified in the fixtures and has become out of sync with next
+        # sequence value
+        # Although I still need to understand if the sequence is being maintained by the database how specifying pkey in
+        # fixtures causes sequence to become out of sync
+        sql = 'SELECT setval(pg_get_serial_sequence(\'advisers_category\',\'id\'), coalesce(max("id"), 1), max("id") IS NOT null) FROM "advisers_category";'
```
Comment (sldblog) [inline laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py]: We tested the category sequence's next value on the environments:

| environment | has a sequence problem |
| --- | --- |
| local development | :warning: yes |
| CI environment (same as local) | :warning: yes |
| staging | :white_check_mark: was fine |
| production | :white_check_mark: was fine |

This command will not run automatically; on local environments and CircleCI the seeded database sequence will still be on "next value = 1" but on staging/production, it is not needed.

In conclusion, this `setval` query only needs to be included to pass local & CircleCI tests. Adding code for tests only works but hides the problem.

I think in order to make the new category additions consistent in all environments, we will need to move it to a migration entry. Categorical data, which cannot be edited by users, can be managed by automated migrations because the outcome is expected to be the same in all environments.

Comment (sldblog) [reply laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py]: What do you think about creating a new migration (and removing this command) that:

1. Fixes the sequence ID by resetting it like the above snippet.
1. Adds the two new categories.

If we do this, the list and IDs will be consistent from development to production.

File: laalaa/apps/advisers/fixtures/initial_advisers.json [hunk 2]
```diff
@@ -86184,7 +86184,8 @@
       2,
       3,
       7,
-      8
+      8,
+      14
```
Comment (sldblog) [inline laalaa/apps/advisers/fixtures/initial_advisers.json]: Why do we need these changes?

File: laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py (hunk 1 above)
Comment (said-moj) [reply laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py]: >1.Fixes the sequence ID by resetting it like the above snippet.
👍 

>Adds the two new categories.
Are we not going to have the same issues we had with cla_backend and the deployment failing due to multiple data migrations runs

Comment (sldblog) [reply laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py]: Maybe, but as the command's code is idempotent now, we could reuse that for the migration too.

File: laalaa/apps/advisers/fixtures/initial_advisers.json (hunk 2 above)
Comment (said-moj) [reply laalaa/apps/advisers/fixtures/initial_advisers.json]: These help setup the initial state. Not the main purpose but It's also useful for making sure there is data for end to end tests

Comment (sldblog) [reply laalaa/apps/advisers/fixtures/initial_advisers.json]: :+1: Gotcha.
Could we expose that intent in a separate commit (and message), if possible, please?

File: laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py (hunk 1 above)
Comment (said-moj) [reply laalaa/apps/advisers/management/commands/add_disc_and_edu_categories.py]: 🎉 Changed command to migration

File: laalaa/apps/advisers/fixtures/initial_advisers.json (hunk 2 above)
Comment (said-moj) [reply laalaa/apps/advisers/fixtures/initial_advisers.json]: 🎉 Added second commit(with message) stating this

File: laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+def create_discrimination_and_education_categories(apps, schema_editor):
+    Category = apps.get_model("advisers", "Category")
+    # Hard coding the id values as this is referenced in advisers/fixtures
+    Category.objects.get_or_create(id=14, code="DISC", defaults={"civil": True})
+    Category.objects.get_or_create(id=15, code="EDU", defaults={"civil": True})
+
+    # This is required because the id field is specified in the fixtures (and above) and has become out of sync with
+    # next sequence value
+    sql = 'SELECT setval(pg_get_serial_sequence(\'advisers_category\',\'id\'), coalesce(max("id"), 1), max("id") IS NOT null) FROM "advisers_category";'
```
Comment (MatMoore) [inline laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py]: I think you can remove the `coalesce` part here, and set the third argument to "false" because you know the table is always populated at the time it will run.

i.e. the following should be equivalent:
`SELECT setval(pg_get_serial_sequence(\'advisers_category\',\'id\'), max("id"), false) FROM "advisers_category"`
or even 
`SELECT setval(pg_get_serial_sequence(\'advisers_category\',\'id\'), 15) FROM "advisers_category"`

which is a bit easier to understand from the context.

File: laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.db import connection
+
+
+def create_discrimination_and_education_categories(apps, schema_editor):
+    Category = apps.get_model("advisers", "Category")
+    # Hard coding the id values as this is referenced in advisers/fixtures
+    Category.objects.get_or_create(id=14, code="DISC", defaults={"civil": True})
```
Comment (MatMoore) [inline laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py]: Does the use of `get_or_create` here mean the environments are not in a consistent state before this migration runs?

Otherwise I would expect you could just run `create` here.

File: laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py (hunk 3 above)
Comment (said-moj) [reply laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py]: That was the output from the `./manage sqlsequencereset` command that prints the SQL statements for resetting sequences for the given app name(s).

File: laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py (hunk 4 above)
Comment (said-moj) [reply laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py]: I think they will be consistent. However we have this strange scenario where our migration runs on the container entrypoint and have decided to make all migrations idempotent

File: laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py (hunk 3 above)
Comment (MatMoore) [reply laalaa/apps/advisers/migrations/0009_adding_discrimination_and_education_categories.py]: Ah ok fair enough, didn't know you could autogenerate that!

File: laalaa/apps/advisers/fixtures/initial_advisers.json (hunk 2 above)
Comment (exonian) [reply laalaa/apps/advisers/fixtures/initial_advisers.json]: Might be worth switching our fixtures to have the code as the natural key at some point (as opposed to using the pks), but not for now.
